### PR TITLE
✨ Implement optional `number` to `str` coercion

### LINKED ISSUE #6045: V2 - Validating int->str
Continue of the #5993

```Python
class Data(BaseModel):
    s: str
    f: str
    d: str


print(Data(s=42, f=4.2, d=Decimal('4.2')))
```

this gives validation error that `s f d - Input should be a valid string`

this was working in pydantic v1, and was actually very handy when dealing with parsing data of a public API where users send messy data from JS client sides

so would be nice to have some feature to be able to turn on str parsing (globally?)

### Affected Components

- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)

**Update**: https://github.com/pydantic/pydantic/issues/6045#issuecomment-1704943954

### PR DESCRIPTION
**Pull request by lig**
## Change Summary

Implement optional `number` to `str` coercion

## Related issue number

Fix #6045 
Requires https://github.com/pydantic/pydantic-core/pull/975

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by lig** [comment]
please review

File: docs/usage/model_config.md [hunk 1]
```diff
@@ -584,6 +585,45 @@ print(type(model2.pet))
...
+class Model(BaseModel):
+    model_config = ConfigDict(coerce_numbers_to_str=True)
+
+    value: str
+
+
+repr(Model(value=42).value)
+#> "42"
+repr(Model(value=42.13).value)
+#> "42.13"
```
Comment (sydney-runkle) [inline docs/usage/model_config.md:623]: Adding an example for `Decimal` as well :)

```suggestion
#> "42.13"
repr(Model(value=Decimal('42.13')).value)
#> "42.13"
```

File: docs/usage/model_config.md [hunk 2]
```diff
@@ -584,6 +585,45 @@ print(type(model2.pet))
...
 
+
+## Coerce Numbers to Strings
+
+Pydantic no longer allows number types (`int`, `float`, `Decimal`) to be coerced as type `str` by default.
+
+Set [`coerce_numbers_to_str=True`](../api/config.md#pydantic.config.ConfigDict.coerce_numbers_to_str) to enable coercing of numbers to strings.
+
+```py
+from pydantic import BaseModel, ConfigDict, ValidationError
```
Comment (sydney-runkle) [inline docs/usage/model_config.md:596]: To support the below `Decimal` example...

```suggestion
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, ValidationError
```

File: docs/usage/model_config.md (hunk 1 above)
Comment (lig) [reply docs/usage/model_config.md:623]: It's another import as well. I didn't want to bloat examples with a lot of boilerplate.

File: docs/usage/model_config.md (hunk 2 above)
Comment (lig) [reply docs/usage/model_config.md:596]: Exactly:)

**Review by sydney-runkle — APPROVED** [review]
LGTM 👍

**Review by adriangb — APPROVED** [review]

**Pull request merged**